File: 2021/Day_02/code.php
<?php

require_once __DIR__ . '/../../lib/php/utils.php';

$input = readInput(__DIR__);

// Task code
function part01(array $input)
{
  $x = $y = 0;
  foreach($input as $v){
    list($dir, $dist) = explode(' ', $v);
    switch ($dir) {
      case 'forward':
        $x += $dist;
        break;
      case 'up':
        $y -= $dist;
        break;
      case 'down':
        $y += $dist;
        break;
      default:
        break;
    }
  }
  return $x * $y;
}

function part02(array $input)
{
  $x = $y = $aim= 0;
  foreach($input as $v){
    list($dir, $dist) = explode(' ', $v);
    switch ($dir) {
      case 'forward':
        $x += $dist;
        $y += $dist * $aim;
        break;
      case 'up':
        $aim -= $dist;
        break;
      case 'down':
        $aim += $dist;
        break;
      default:
        break;
    }
  }
  return $x * $y;
}

// Execute
calcExecutionTime();
$result01 = part01($input);
$result02 = part02($input);
$executionTime = calcExecutionTime();

writeln('Solution Part 1: ' . $result01);
writeln('Solution Part 2: ' . $result02);
writeln('Execution time: ' . $executionTime);

saveBenchmarkTime($executionTime, __DIR__);

// Task test
testResults(
    [], // Expected
    [$result01, $result02], // Result
);

File: 2021/Day_02/code.go
package main

import (
	"advent-of-code/helper"
	"fmt"
	"path"
	"runtime"
	"strconv"
	"strings"
	"time"
)

func main() {
	_, f, _, _ := runtime.Caller(0)
	cwd := path.Join(path.Dir(f))

	input := helper.ReadInput(cwd, helper.NewLine)
	iValues := input.Strings()

	// Execute
	start := time.Now()
	result01 := part01(iValues)
	result02 := part02(iValues)
	executionTime := helper.ExecutionTime(time.Since(start))

	fmt.Printf("Solution Part 1: %v\n", result01)
	fmt.Printf("Solution Part 2: %v\n", result02)
	fmt.Printf("Execution time: %s\n", executionTime)

	helper.SaveBenchmarkTime(executionTime, cwd)

	// Testing
	helper.TestResults(
		[]helper.TestingValue{
			helper.TestingValue{Result: result01, Expect: 1561344},
			helper.TestingValue{Result: result02, Expect: 1848454425},
		},
	)
}

// Task code
func part01(input []string) int {
	x, y := 0, 0
	for _, v := range input {
		s := strings.Split(v, " ")
		n, _ := strconv.Atoi(s[1])
		switch s[0] {
		case "forward":
			x += n
			break
		case "up":
			y -= n
			break
		case "down":
			y += n
			break
		}
	}
	return x * y
}

func part02(input []string) int {
	x, y, aim := 0, 0, 0
	for _, v := range input {
		s := strings.Split(v, " ")
		n, _ := strconv.Atoi(s[1])
		switch s[0] {
		case "forward":
			x += n
			y += n * aim
			break
		case "up":
			aim -= n
			break
		case "down":
			aim += n
			break
		}
	}
	return x * y
}
